src/Features/LanguageServer/ProtocolUnitTests/Workspaces/LspWorkspaceManagerTests.cs

```diff
@@ -480,6 +531,104 @@ public async Task TestDoesNotForkWhenDocumentTextBufferOpenedAsync()
+
+        // Open the doc
+        var filePath = @"c:\Test1.cs";
+        var documentUri = new Uri("file://" + filePath, UriKind.Absolute);
+        await testLspServer.OpenDocumentAsync(documentUri, "Text");
+
+        // Initially the doc will be in the lsp misc workspace.
+        var (workspace1, document1) = await GetLspWorkspaceAndDocumentAsync(documentUri, testLspServer).ConfigureAwait(false);
+        Assert.Equal(WorkspaceKind.MiscellaneousFiles, workspace1?.Kind);
+
+        AssertEx.NotNull(document1);
+        Assert.Equal("Text", (await document1.GetTextAsync(CancellationToken.None)).ToString());
+
+        // Now, add a project to the actual workspace containing Test1.cs with different content.
+        await testLspServer.TestWorkspace.ChangeSolutionAsync(
+            testLspServer.TestWorkspace.CurrentSolution.Projects.Single()
+                .AddDocument(filePath, SourceText.From("ProjectSystemText"), filePath: filePath)
+                .Project.Solution);
+
+        // if we were to directly query the workspace right now, the contents won't match what lsp thinks it is:
+        Assert.NotEqual(
+            (await document1.GetTextAsync(CancellationToken.None)).ToString(),
+            (await testLspServer.TestWorkspace.CurrentSolution.Projects.Single().Documents.Single().GetTextAsync()).ToString());
+
+        // However, if we retrieve from lsp now, we'll see the document moved into the real workspace and we'll see the
+        // real workspace contents changed.
+        (workspace1, document1) = await GetLspWorkspaceAndDocumentAsync(documentUri, testLspServer).ConfigureAwait(false);
+        Assert.Equal(WorkspaceKind.Host, workspace1?.Kind);
+
+        AssertEx.NotNull(document1);
+        Assert.Equal("Text", (await document1.GetTextAsync(CancellationToken.None)).ToString());
+
+        // Retrieving the document from lsp should push its content into the workspace.
+        Assert.Equal(
+            (await document1.GetTextAsync(CancellationToken.None)).ToString(),
+            (await testLspServer.TestWorkspace.CurrentSolution.Projects.Single().Documents.Single().GetTextAsync()).ToString());
+
+        // The file should not be in the misc workspace.
+        Assert.Empty(testLspServer.GetManagerAccessor().GetLspMiscellaneousFilesWorkspace()!.CurrentSolution.Projects);
+
+        // Now, if the project system removes the file, we will still see the lsp version of, but back to the misc workspace.
+        await testLspServer.TestWorkspace.ChangeSolutionAsync(
+            testLspServer.TestWorkspace.CurrentSolution.Projects.Single().RemoveDocument(document1.Id).Solution);
+
+        (workspace1, document1) = await GetLspWorkspaceAndDocumentAsync(documentUri, testLspServer).ConfigureAwait(false);
+        Assert.Equal(WorkspaceKind.MiscellaneousFiles, workspace1?.Kind);
+
+        AssertEx.NotNull(document1);
+        Assert.Equal("Text", (await document1.GetTextAsync(CancellationToken.None)).ToString());
```

Comment:
great test 🥳 !

Reply:
Indeed, this set of tests are fantastic.